consensus/src/experimental/commit_phase.rs

```diff
@@ -0,0 +1,333 @@
+        )
+        .await
+}
+
+impl CommitPhase {
+    pub fn new(
+        commit_channel_recv: Receiver<(Vec<ExecutedBlock>, LedgerInfoWithSignatures)>,
+        execution_proxy: Arc<dyn StateComputer>,
+        commit_msg_sender: channel::Sender<CommitPhaseChannelMsgWrapper>,
+        commit_msg_receiver: channel::Receiver<ConsensusMsg>,
+        verifier: ValidatorVerifier,
+        safety_rules: Arc<Mutex<MetricsSafetyRules>>,
+        author: Author,
+    ) -> Self {
+        Self {
+            commit_channel_recv,
+            execution_proxy,
+            local_cache: HashMap::<HashValue, LedgerInfoWithSignatures>::new(),
+            local_queue: VecDeque::<CacheItem>::new(),
+            commit_msg_sender,
+            commit_msg_receiver,
+            verifier,
+            safety_rules,
+            author,
+            committed_round: 0,
+        }
+    }
+
+    /// Notified when receiving a commit vote message
+    pub async fn process_commit_vote(&mut self, commit_vote: &CommitVote) -> anyhow::Result<()> {
+        let li = commit_vote.ledger_info();
+
+        if li.commit_info().round() < self.committed_round {
+            return Ok(()); // we ignore the message
+        }
+
+        // verify the signature
+        commit_vote.verify(&self.verifier)?;
+
+        let chash = li.commit_info().executed_state_id();
+
+        // add the signature to local_cache
+        match self.local_cache.entry(chash) {
+            Entry::Occupied(mut cio) => {
+                let ci = cio.get_mut();
+                ci.add_signature(commit_vote.author(), commit_vote.signature().clone());
+            }
+            Entry::Vacant(_) => {
+                let mut li_sig = LedgerInfoWithSignatures::new(
+                    li.clone(),
+                    BTreeMap::<AccountAddress, Ed25519Signature>::new(),
+                );
+                li_sig.add_signature(commit_vote.author(), commit_vote.signature().clone());
+                self.local_cache.insert(chash, li_sig);
+            }
+        };
+
+        Ok(())
+    }
+
+    pub async fn process_commit_decision(
+        &mut self,
+        commit_decision: CommitDecision,
+    ) -> anyhow::Result<()> {
+        let li = commit_decision.ledger_info();
+
+        if li.ledger_info().commit_info().round() < self.committed_round {
+            return Ok(()); // we ignore the message
+        }
+
+        commit_decision.verify(&self.verifier)?;
+
+        let chash = li.ledger_info().commit_info().executed_state_id();
+
+        // TODO: optimization1: probe local_cache first to see if the existing already verifies,
+        // TODO: otherwise we do not make changes.
+        // TODO: optimization2: we can set a bit to indicate this tree of signatures are already verified,
+        // TODO: we do not have to verify it again in the main loop.
+
+        // replace the signature tree directly if it does not verify
+        self.local_cache.insert(chash, li.clone());
```

Comment:
we need to think about how to handle unhappy path, this cache could potentially go out of memory if nodes are way behind (keep receiving about something far in the future).

this is like passive sync up, and we do active sync up in consensus (once we receive something in the future we directly sync to)